### DISCUSSION TITLE
When using JWT, adapter not called

### DISCUSSION THREAD

**Discussion by y-nk** [nextauthjs/next-auth#7490] [Help]
### Environment

next: 13.4
next-auth: 4.22.1
prisma-adapter: 1.0.6

### Reproduction URL

https://example.com/

### Describe the issue

I'm running a standard Next.JS 13 setup with:

- prisma adapter
- email provider
- google provider

I've added a callback on `session` to console.log the data in the session.

My findings are:

- without any options, the `session.user` matches the one inserted in the database
- when i add `session { strategy: 'jwt' }`, the `session.user` is constantly undefined, for both providers.

For every test i did:

1. empty the cookies
2. empty the database (to avoid link account issues)
3. proceed to the auth flow
4. redirect to home and log the content of `getServerSession`

### How to reproduce

It's hard to provide a setup since i'd need either a google oauth client/secret pair and a smtp server.

### Expected behavior

I think the `session.user` should be fetched from the database all the time, even when the `session.strategy = 'jwt'`

**Comment by y-nk** [comment]
### Update
I just found this information [in a code comment](https://next-auth.js.org/configuration/options#session) (not so obvious, but i should have read it...):

> ```js
> // If you use an `adapter` however, we default it to `"database"` instead.
> // You can still force a JWT session by explicitly defining `"jwt"`.
> ```

I'd like to update my question: I'd love to keep using jwt tokens, but ultimately i would like the database user object to be populated in the session.

**How do I achieve that?**

**Comment by y-nk** [comment]
**While waiting for a better answer** the workaround is to do manually a pull from the database in the jwt callback and decorate the session with it in the session callback.

```js
callbacks: {
  async jwt({ token }) {
    if (token.email) {
      token.user = await prisma.user.findUnique({
        where: {
          email: token.email
        }
      })
    }
  
    return token
  },
  
  async session({ session, token }) {
    if (token.user) {
      session.user = token.user
    }
  
    return session
  },
}
```

I'm not fond of this solution because the `token.user` is typed as `unknown` and it looks like there's nothing we can do about it. The type isn't propagated to the `session` callback neither therefore the type of `session.user` is incorrect in the returned session.

Hopefully, we can still pass the session type in the `getServerSession` like this:

```ts
import type { Session } from 'next-auth';
import { getServerSession } from 'next-auth';

import { authOptions } from './authConfig';

import type { User } from '@prisma/client';

type AppSession = Session & {
  user: User
}

export async function getUser() {
  const session = await getServerSession<typeof authOptions, AppSession>(authOptions)
  return session?.user
}
```

it works because the types are shallow at the moment, but not sure on how it should be "the proper way"

**Comment by goerlitz** [comment]
I'm running into the same issue.

I'm using a Prisma adapter and middleware which makes things really complicated.

1. `If you use an adapter however, we default it to "database" instead.` -> this makes the middleware redirect infinitely :(
2. Setting session back to `session: { strategy: 'jwt' }` fixes the middleware issue but does not populate user in session callback as described above
3. Not having user populated in session leads to more issues, e.g. the documentation on [implementing role-based access](https://authjs.dev/guides/basics/role-based-access-control#with-database) is misleading - instead of following `with database` I need to follow `with JWT`. However, the first sentence of `with JWT` says "When you don't have a database configured" (doh).

**UPDATE**: I found a solution that works for me.

`authOptions`:
```ts
callbacks: {
  jwt({ token, user }) {
    // set user role in jwt (executed before session callback)
    if (user) token.role = user.role;
    return token;
  },
  session({ session, token }) {
    // set user role in session for role based access
    session.user.role = token.role as string;
    return session;
  }
}
```
and in `types/next-auth.d.ts`:
```ts
// see https://next-auth.js.org/getting-started/typescript#module-augmentation
import { DefaultSession } from 'next-auth';

declare module 'next-auth' {
  interface User {
    role?: string;
  }

  /**
   * Returned by `useSession`, `getSession` and received as a prop on the `SessionProvider` React Context
   */
  interface Session {
    user: User & DefaultSession['user'];
  }
}
```

**Comment by parthy007** [comment]
I'm using mongoDB adapter but with this I'm able to login with the google provider but not with my credentials, now in order to login with my credentials I had to do session:{strategy:"jwt"} which I guess looks for session in the cookie forcefully even with the adapter. Now this again hinders with the google provider as it looks for session in the database. How to go around this, I'm new to Nextjs so couldn't really understand. The above discussion is the closest to my problem so would appreciate any help.


```import { User } from "@/models/User";
import mongoose from "mongoose";
import NextAuth from "next-auth";
import CredentialsProvider from "next-auth/providers/credentials";
import GoogleProvider from "next-auth/providers/google";
import bcrypt from "bcrypt";
import { MongoDBAdapter } from "@auth/mongodb-adapter";
import clientPromise from "@/libs/mongoConnect";
import { getServerSession } from "next-auth";

async function getUser() {
  const session = await getServerSession(authOptions);
  return session?.user;
}

export const authOptions = {
  adapter: MongoDBAdapter(clientPromise),
  secret: process.env.SECRET,
  providers: [
    GoogleProvider({
      id: "google",
      clientId: process.env.GOOGLE_CLIENT_ID,
      clientSecret: process.env.GOOGLE_CLIENT_SECRET,
    }),
    CredentialsProvider({
      name: "Credentials",
      id: "credentials",
      credentials: {
        username: {
          label: "Email",
          type: "email",
          placeholder: "test@example.com",
        },
        password: { label: "Password", type: "password" },
      },
      async authorize(credentials, req) {
        const email = credentials?.email;
        const password = credentials?.password;

        mongoose.connect(process.env.MONGO_URL);
        const user = await User.findOne({ email });
        const passwordOk = user && bcrypt.compareSync(password, user.password);

        if (passwordOk) {
          return user;
        }

        return null;
      },
    }),
  ],
};

const handler = NextAuth(authOptions);

export { handler as GET, handler as POST };
```
